tokio/src/sync/set_once.rs

```diff
@@ -0,0 +1,402 @@
+        if self.initialized() {
+            return Err(SetOnceError(value));
+        }
+
+        // SAFETY: lock the mutex to ensure only one caller of set
+        // can run at a time.
+        let guard = self.lock.lock();
+
+        if self.initialized() {
+            // If the value is already set, we return an error
+            drop(guard);
+
+            return Err(SetOnceError(value));
+        }
+
+        // SAFETY: We have locked the mutex and checked if the value is
+        // initalized or not, so we can safely write to the value
+        unsafe {
+            self.value.with_mut(|ptr| (*ptr).as_mut_ptr().write(value));
+        }
+
+        // Using release ordering so any threads that read a true from this
+        // atomic is able to read the value we just stored.
+        self.value_set.store(true, Ordering::Release);
+
+        drop(guard);
+
+        // notify the waiting wakers that the value is set
+        self.notify.notify_waiters();
+
+        Ok(())
+    }
+
+    /// Takes the value from the cell, destroying the cell in the process.
+    /// Returns `None` if the cell is empty.
+    pub fn into_inner(self) -> Option<T> {
+        // TODO: Use get_mut()
+        let value_set = self.value_set.load(Ordering::Relaxed);
+
+        if value_set {
+            // Since we have taken ownership of self, its drop implementation
+            // will be called by the end of this function, to prevent a double
+            // free we will set the value_set to false so that the drop
+            // implementation does not try to drop the value again.
+            self.value_set.store(false, Ordering::Relaxed);
+
+            // SAFETY: The SetOnce is currently initialized, we can assume the
+            // value is initialized and return that, when we return the value
+            // we give the drop handler to the return scope.
+            Some(unsafe { self.value.with_mut(|ptr| ptr::read(ptr).assume_init()) })
+        } else {
+            None
+        }
+    }
+
+    /// Waits until set is called. The future returned will keep blocking until
+    /// the `SetOnce` is initialized.
+    ///
+    /// If the `SetOnce` is already initialized, it will return the value
+    /// immediately.
+    ///
+    /// # Note
+    ///
+    /// This will keep waiting until the `SetOnce` is initialized, so it
+    /// should be used with care to avoid blocking the current task
+    /// indefinitely.
+    pub async fn wait(&self) -> &T {
+        loop {
+            if let Some(val) = self.get() {
+                return val;
+            }
+
+            let notify_fut = self.notify.notified();
+            // take the lock because we're reading value_set
+            let guard = self.lock.lock();
+
+            if self.value_set.load(Ordering::Relaxed) {
+                // SAFETY: the state is initialized
+                return unsafe { self.get_unchecked() };
+            }
+
+            drop(guard);
```

Comment:
I want to add a scope here for the guard. There are some cases where using `drop(guard)` on a non-Send value causes Rust to still think that the future is non-Send when it should be Send. It looks like that may not be the case here as `async_send_sync.rs` seems to be okay, but I feel better with the scope.

I also took the opportunity to improve the comment.
```suggestion
            {
                // Taking the lock here ensures that a concurrent call to `set`
                // will see the creation of `notify_fut` in case the check
                // fails.
                let guard = self.lock.lock();

                if self.value_set.load(Ordering::Relaxed) {
                    // SAFETY: the state is initialized
                    return unsafe { self.get_unchecked() };
                }
            }
```